src/synapse/api.py

```diff
@@ -0,0 +1,459 @@
+    """Register user.
+
+    Args:
+        registration_shared_secret: secret to be used to register the user.
+        user: user to be registered.
+        server: to be used to create the user id.
+        admin_access_token: admin access token to get user's access token if it exists.
+
+    Raises:
+        RegisterUserError: if there was an error registering the user.
+
+    Returns:
+        Access token to be used by the user.
+    """
+    # get nonce
+    nonce = _get_nonce()
+    # generate mac
+    hex_mac = _generate_mac(
+        shared_secret=registration_shared_secret,
+        nonce=nonce,
+        user=user.username,
+        password=user.password,
+        admin=user.admin,
+    )
+    # build data
+    data = {
+        "nonce": nonce,
+        "username": user.username,
+        "password": user.password,
+        "mac": hex_mac,
+        "admin": user.admin,
+    }
+    # finally register user
+    try:
+        res = _do_request("POST", REGISTER_URL, json=data)
+        return res.json()["access_token"]
+    except UserExistsError as exc:
+        logger.warning("User %s exists, getting the access token.", user.username)
+        if not server or not admin_access_token:
+            raise RegisterUserError(
+                f"User {user.username} exists but there is no server/admin access token set."
+            ) from exc
+        return get_access_token(
+            user=user, server=str(server), admin_access_token=str(admin_access_token)
+        )
+    except (requests.exceptions.JSONDecodeError, TypeError, KeyError) as exc:
+        logger.exception("Failed to decode access_token: %r. Received: %s", exc, res.text)
+        raise RegisterUserError(str(exc)) from exc
+
+
+def _generate_mac(
+    shared_secret: str,
+    nonce: str,
+    user: str,
+    password: str,
+    admin: str | bool = False,
+    user_type: typing.Optional[str] = None,
+) -> str:
+    """Generate mac to register user.
+
+    "The MAC is the hex digest output of the HMAC-SHA1 algorithm, with the key being the shared
+    secret and the content being the nonce, user, password, either the string "admin" or
+    "notadmin", and optionally the user_type each separated by NULs.".
+    Extracted from: https://matrix-org.github.io/synapse/latest/admin_api/register_api.html
+
+    Args:
+        shared_secret: registration_shared_secret from configuration file.
+        nonce: nonce generated.
+        user: username for the new user.
+        password: password used for authentication.
+        admin: if is admin. Default False.
+        user_type: user type. Defaults to None.
+
+    Returns:
+        User in HMAC format as a string of hexadecimals.
+        This format is expected by the Synapse API.
+    """
+    mac = hmac.new(key=shared_secret.encode("utf8"), digestmod=hashlib.sha1)
+    mac.update(nonce.encode("utf8"))
+    mac.update(b"\x00")
+    mac.update(user.encode("utf8"))
+    mac.update(b"\x00")
+    mac.update(password.encode("utf8"))
```

Comment:
Since the password is hashed into this mac and it is using SHA1 we should check whether we log the mac anywhere - it should essentially be treated the same as the password in terms of sensitivity

As a side note, this seems to be quite a risky implementation on Synapse's part, the reason for using an unsafe hashing algorithm is not clear and putting the password into it is just waiting for someone to log the mac and cause a security incident

Reply:
Also - it might be worth looking into how Synapse uses this mac - if it is stored anywhere or logged anywhere, for example, that would be highly risky

Reply:
This might be something to file an issue on the Synapse project.

For reference this method is provided by the documentation [here](https://matrix-org.github.io/synapse/latest/admin_api/register_api.html?highlight=mac#shared-secret-registration) and is supposed to be deprecated.

Reply:
Yeh I agree we should raise it as an issue with Synapse. Is there an alternative since this method seems to be deprecated?